homeassistant/components/picnic/services.py

```diff
@@ -0,0 +1,175 @@
+                }
+            ),
+        )
+
+    # Register the search service
+    if not hass.services.has_service(DOMAIN, SERVICE_SEARCH_PRODUCT):
+
+        async def async_search_product_service(call: ServiceCall):
+            # Get the picnic API client and call the handler with api client and hass
+            api_client: PicnicAPI = await get_api_client(
+                hass, call.data.get("device_id")
+            )
+            await handle_search(hass, api_client, call)
+
+        hass.services.async_register(
+            DOMAIN,
+            SERVICE_SEARCH_PRODUCT,
+            async_search_product_service,
+            schema=vol.Schema(
+                {
+                    vol.Optional(ATTR_DEVICE_ID): cv.string,
+                    vol.Required(ATTR_PRODUCT_NAME): cv.string,
+                }
+            ),
+        )
+
+
+async def get_api_client(hass, device_id: str = None):
+    """Get the right Picnic API client based on the device id, else get the default one."""
+    if device_id is None:
+        default_config_id = list(hass.data[DOMAIN].keys())[0]
+        return hass.data[DOMAIN][default_config_id][CONF_API]
+
+    # Get device from registry
+    registry = await device_registry.async_get_registry(hass)
+    device = registry.async_get(device_id)
+
+    if device is None:
+        raise PicnicServiceException(f"Device with id {device_id} not found!")
+
+    # Get Picnic API client for the config entry id
+    try:
+        config_entry_id = next(iter(device.config_entries))
+        return hass.data[DOMAIN][config_entry_id][CONF_API]
+    except (AttributeError, StopIteration, KeyError) as error:
+        raise PicnicServiceException(
+            f"Device with id {device_id} not found!"
+        ) from error
+
+
+async def handle_add_product(
+    hass: HomeAssistant, api_client: PicnicAPI, call: ServiceCall
+) -> None:
+    """Handle the call for the add_product service."""
+    product_id = call.data.get("product_id")
+    if not product_id:
+        search_results = await hass.async_add_executor_job(
+            _product_search, api_client, call.data.get("product_name")
+        )
+        try:
+            product_id = search_results[0]["id"]
+        except (IndexError, KeyError, TypeError):
+            product_id = None
+
+    if not product_id:
+        raise PicnicServiceException("No product found or no product ID given!")
+
+    await hass.async_add_executor_job(
+        api_client.add_product, str(product_id), call.data.get("amount", 1)
+    )
+
+
+async def handle_search(
+    hass: HomeAssistant, api_client: PicnicAPI, call: ServiceCall
+) -> None:
+    """Handle the call for the search service."""
+    products = await hass.async_add_executor_job(
+        _product_search, api_client, call.data.get("product_name")
+    )
+
+    products_dict = {p["id"]: p for p in products[:5]}
+    hass.bus.async_fire("picnic_search_result", products_dict)
```

Comment:
Smart!

But not a construct we want to use. Right now, we don't have any constructs for command with results in Home Assistant. So, unfortunately, at this point in time, you can't build this.

Reply:
Ok I suspected this wouldn't be something we'd want to support with out a proper construction.

Btw I am exploring this and had a proof of concept working, but have a lot to do to make a proposal.

Reply:
Ok, I'll remove it!